[ingestion] webhooks handling

jerome-poisson opened the pull request:
this commit add a basic handling of webhooks.
For now settings is done on backend using FEEDING_WEBHOOKS where key is
provider name and value is endpoint suffix.

when adding a webhook, e.g. with FEEDING_WEBHOOKS = {'Personalia':
'wufoo'}, a webhook_[endpoint_suffix] will be used (the prefix help to
prevent conflicts), so here webhook_wufoo will be used.

When an GET or POST request is done on webhook_wufoo, this will trigger
the Personalia provider update.

This is a first step, improvments like client-side configuration and
new option in update schedule could follow.

This commit also add handling of transparent_schema_rules in Resource.

SDNTB-378

jerome-poisson commented:
This P.R. hasn't been tested with Wufoo yet, it needs https://github.com/superdesk/superdesk-core/pull/799 to be merged, and a publicly available URL.

ioanpocol commented:
please rebase

petrjasek commented on superdesk/default_settings.py:
```diff
@@ -571,5 +572,12 @@ def celery_queue(name):
+#: dict were key is provider name and value is webhook endpoint
+#: "webhook_" will always be prefixed to endpoint name to avoid name conflicts
+#: e.g. {'wufoo NTB': 'wufoo'} will create webhook_wufoo endpoint
+FEEDING_WEBHOOKS = {}
```
imo we should introduce an api for this, like we have to register ingest providers, outputs etc.

petrjasek commented on apps/io/feeding_service_webhook.py:
```diff
@@ -0,0 +1,58 @@
+
+
+class FeedingServiceWebhookResource(Resource):
+    public_methods = ['GET', 'POST']
```
there is missing authentication imo. POST shouldn't be public, and not sure if there is a usecase for GET requests?

jerome-poisson replied:
@petrjasek this endpoint is there for being called by external services, it's just a generic public URL which will trigger the feeding service every time a POST or GET request is made there. It's public and without authentication on purpose.

petrjasek replied:
which is imo wrong, because that makes it vulnerable to dos attacks

jerome-poisson replied:
@petrjasek URL must be generic and is called by third party services, we can't use authentication there because we have no control on the entity doing the HTTP request. In the case of Wufoo, they have a handshake parameter which we could use, but this would be vendor specific.

Furthermore, the risk is limited, as the attacker need to know an URL, and with it he can only trigger a ingestion, there are probably more efficient ways to attack a superdesk instance.

We can add protections like limiting the number of webhook triggers (no more than 1 every X minutes for instance), or restrict to a specific request IP.

This P.R. only add a simple way to trigger ingestion with webhook, but if the security is a concern here, we should discuss it more deeply to decide how to implement the whole thing.

The POST doesn't actually store anything, there is no backend attached to the service.

petrjasek replied:
if we control the url, we can provide auth tokens for it right? or at least add some limit.
besides this one there is only auth endpoint which is public and that one requires credentials, so not sure what you mean by more efficient ways.

naspeh replied:
usually it's good practice to have some kind of  `Signature` for example: https://developer.github.com/webhooks/#delivery-headers
```
X-Hub-Signature	HMAC hex digest of the payload, using the hook's secret as the key (if configured).
```
that you can identify that the posted request is part that has secret key...

I don't understand why Wufoo don't use it, this `a handshake parameter` I think is just `GET` parameter like secret, but the secret shouldn't be transferred over internet...

so in case of Wufoo probably IP restriction should be used or so.

naspeh replied:
or maybe really use this `a handshake parameter` but HTTPS should be used for sure in this case...

petrjasek replied:
imo we can't assume what this entity making post requests will implement, but we control the url..

jerome-poisson replied:
@petrjasek I'm not sure what you mean with auth token, the requesting entity (wufoo) will not provide it. By "more efficient way", I mean that if somebody really want to do a DoS, he'll better go with a DDoS tool like LOIC than guessing a specific url (which is not supposed to be public, even if for NTB it's currently hardcoded) to trigger ingestion.

The issue by using handshake or other shared secret is that it is vendor specific and not generic anymore. A hidden URL is a kind of shared secret anyway (we can autogenerate a key in the URL). An yes HTTPS is needed.

Maybe we can talk about this during the scrum?

petrjasek replied:
for the record, what I mean is that we generate url for different push services like:
`superdesk.org/api/push?auth_token=abc`

jerome-poisson commented:
@petrjasek @naspeh after this morning discussion, I've added an authentication mechanism based on auth parameter and environment variable.
I'm not really satisfied with the hack done for the Service because no backend is attached, I'm not use if there is a simpler way to do that with Eve (maybe adding an URL directly with flask?). We can discuss it tomorrow.

Also @naspeh, I would like to set the environment variable for the test instance to try it with Wufoo, if we can check that together tomorrow, thanks :)

petrjasek commented on apps/io/feeding_service_webhook.py:
```diff
@@ -0,0 +1,100 @@
+log = logging.getLogger(__name__)
+
+
+class FeedinServviceWebhookAuth(object):
```
typo Ser**vv**ice

jerome-poisson replied:
right thanks

petrjasek commented on apps/io/feeding_service_webhook.py:
```diff
@@ -0,0 +1,100 @@
+
+
+class FeedingServiceWebhookResource(Resource):
+    resource_methods = ['GET', 'POST']
```
I would remove `GET` if possible, imo it triggers an action, so should be `POST`

jerome-poisson replied:
I've put it just in case, but we can still add it in the future if needed, I'll remove

petrjasek commented on superdesk/io/registry.py:
```diff
@@ -115,3 +118,29 @@ def service(service_id):
         )
+
+
+def register_webhook(provider_name, webhook_name, webhook_service=None):
```
actually why would we create multiple resources for this? couldn't be provider a param as well?
`/webhook?provider=woofoo`

jerome-poisson replied:
yes that's an option which is probably cleaner, I'll change that

petrjasek replied:
maybe this should be provider id rather than type, it might make things simpler

jerome-poisson replied:
@petrjasek provider id is not visible in ingest setting UI. As we have to construct URL manually, I'll use provider name for now.
In a future step, when we change the UI, I think a webhook URL should be automatically generated (with auth key) and displayed, and provider id should then be used indeed.

naspeh commented on apps/io/feeding_service_webhook.py:
```diff
@@ -0,0 +1,100 @@
+        abort(403, description='You are not authorized to access this resource')
+
+    def authorized(self, allowed_roles, resource, method):
+        env_name = '{}_auth'.format(resource)
```
probably name should be in uppercase format

jerome-poisson replied:
I was hesitating actually :). . OK I'll change

jerome-poisson commented:
@petrjasek @naspeh I've refactored the patch and removed multiple endpoints and GET handling. As a result the P.R. is much simpler, and there is no more need for a registration API (variable environment is enough). I'll close the P.R. on ntb repos which is no more needed. I've also improved the doc.

jerome-poisson commented:
tested OK with Wufoo

petrjasek commented on superdesk/io/webhooks/__init__.py:
```diff
@@ -0,0 +1,77 @@
+    def authorized(self, allowed_roles, resource, method):
+        try:
+            # provider's id may be used in the future
+            provider_name = request.args['provider_name']
```
imo this part shouldn't be in authorization handling, only later.

jerome-poisson replied:
@petrjasek that's because there is a bug, I've renamed the variable but forgot to put it in the lookup, it should be used to start only the provider requested, and for now it's starting all providers, I'll fix that

jerome-poisson replied:
@petrjasek actually forget my last message, I've misread the diff and the lookup is OK.
the provider_name is needed here for checking the environment variable, and the page is not authorised if nothing is set in environment, so that's kind of needed here.

It may still be changed later when we have some client UI, because at this time the environment variable should not be used anymore.
